src/compiler/builder.ts

```diff
@@ -0,0 +1,528 @@
+                    if (symbol && symbol.declarations && symbol.declarations[0]) {
+                        const declarationSourceFile = getSourceFileOfNode(symbol.declarations[0]);
+                        if (declarationSourceFile) {
+                            addReferencedFile(declarationSourceFile.path);
+                        }
+                    }
+                }
+            }
+
+            const sourceFileDirectory = getDirectoryPath(sourceFile.path);
+            // Handle triple slash references
+            if (sourceFile.referencedFiles && sourceFile.referencedFiles.length > 0) {
+                for (const referencedFile of sourceFile.referencedFiles) {
+                    const referencedPath = toPath(referencedFile.fileName, sourceFileDirectory, getCanonicalFileName);
+                    addReferencedFile(referencedPath);
+                }
+            }
+
+            // Handle type reference directives
+            if (sourceFile.resolvedTypeReferenceDirectiveNames) {
+                sourceFile.resolvedTypeReferenceDirectiveNames.forEach((resolvedTypeReferenceDirective) => {
+                    if (!resolvedTypeReferenceDirective) {
+                        return;
+                    }
+
+                    const fileName = resolvedTypeReferenceDirective.resolvedFileName;
+                    const typeFilePath = toPath(fileName, sourceFileDirectory, getCanonicalFileName);
+                    addReferencedFile(typeFilePath);
+                });
+            }
+
+            return referencedFiles;
+
+            function addReferencedFile(referencedPath: Path) {
+                if (!referencedFiles) {
+                    referencedFiles = createMap<true>();
+                }
+                referencedFiles.set(referencedPath, true);
+            }
+        }
+
+        /**
+         * Gets all the emittable files from the program
+         */
+        function getAllEmittableFiles(program: Program) {
+            const defaultLibraryFileName = getDefaultLibFileName(program.getCompilerOptions());
+            const sourceFiles = program.getSourceFiles();
+            const result: string[] = [];
+            for (const sourceFile of sourceFiles) {
+                if (getBaseFileName(sourceFile.fileName) !== defaultLibraryFileName && shouldEmitFile(sourceFile)) {
+                    result.push(sourceFile.fileName);
+                }
+            }
+            return result;
+        }
+
+        function getNonModuleEmitHandler(): EmitHandler {
+            return {
+                onAddSourceFile: noop,
+                onRemoveSourceFile: noop,
+                onUpdateSourceFile: noop,
+                onUpdateSourceFileWithSameVersion: returnFalse,
+                getFilesAffectedByUpdatedShape
+            };
+
+            function getFilesAffectedByUpdatedShape(program: Program, _sourceFile: SourceFile, singleFileResult: string[]): string[] {

```

Comment:
It looks like this would be simpler if it took and returned `SourceFile[]` instead of `string[]`.
Then you can call `onAffectedFile(file.fileName, file);` instead of `onAffectedFile(file, program.getSourceFile(file));`
(Also, since `file.fileName` can trivially be derived from `file`, you could skip passing that into the callback too.)
You might similarly change the return type of `getFilesAffectedBy`. to `SourceFile[]`.

Reply:
getFilesAffectedByUpdatedShape is used by project to get affected file names, so this works out better than sending source files so that project doesnt need to iterate on the list to convert that to file names.. In rest of the cases, you can always get the source file from file name

Reply:
But, it's simpler to get the filename from a source file (`sf.fileName`) than it is to get the source file from a file name (requires a lookup table).